uberfire-nio2-backport/uberfire-nio2-impls/uberfire-nio2-jgit/src/main/java/org/uberfire/java/nio/fs/jgit/manager/JGitFileSystemsCacheDataStructure.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2018 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.uberfire.java.nio.fs.jgit.manager;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+
+import org.uberfire.java.nio.fs.jgit.JGitFileSystem;
+import org.uberfire.java.nio.fs.jgit.JGitFileSystemProviderConfiguration;
+
+public class JGitFileSystemsCacheDataStructure {
+
+    public static Map<String, Supplier<JGitFileSystem>> create(JGitFileSystemProviderConfiguration config) {
+
+        return Collections.synchronizedMap(new LinkedHashMap<String, Supplier<JGitFileSystem>>(config.getJgitFileSystemsInstancesCache() + 1,
+                                                                                               0.75f,
+                                                                                               true) {
+
+            //prevent infinite loop if all FS instances are in use
+            private Integer removeEldestEntryIterations = 0;
+
+            @Override
+            public Supplier<JGitFileSystem> putIfAbsent(String key, Supplier<JGitFileSystem> value) {
+                Supplier<JGitFileSystem> jGitFileSystemSupplier = super.putIfAbsent(key, value);
+                if (size() > config.getJgitFileSystemsInstancesCache()) {
+                    fitListToCacheSize();
+                }
+                return jGitFileSystemSupplier;
+            }
+
+            private void fitListToCacheSize() {
+                List<String> itemsToRemove = new ArrayList<>();
+                int maxIterations = config.getJgitCacheOverflowCleanupSize();
+                for (int i = this.size() - 1; (i >= 0 && (this.size() - i < maxIterations)); i--) {
+                    Map.Entry<String, Supplier<JGitFileSystem>> entry = (Map.Entry) this.entrySet().toArray()[i];
+                    JGitFileSystem targetFS = (JGitFileSystem) ((MemoizedFileSystemsSupplier) entry.getValue()).get();
+                    if (!targetFS.isInUse()) {
+                        itemsToRemove.add(entry.getKey());
+                    }
+                }
+                if (!itemsToRemove.isEmpty()) {
+                    itemsToRemove.forEach(item -> this.remove(item));
```

Comment:
as the action of inUse verification and the exclusion are not atomic, could you please make sure that we have a test that shows that it would be ok to have 2 instances of the file system.
as from line 54 to line 59 the instance could be `InUse` and still removed from cache (what should be fine, we just need a test to verify it).

test should be simple as add one FS to the cache, remove it from cache while holding an instance of the object and them make some operaiton followed by another hit in the cache that would return a new instance, just check if both locks are the same and all good! ;)